src/SignalR/clients/ts/signalr/src/HubConnection.ts

```diff
@@ -156,13 +156,16 @@ export class HubConnection {
         const [streams, streamIds] = this.replaceStreamingParams(args);
         const invocationDescriptor = this.createStreamInvocation(methodName, args, streamIds);
 
+        let promiseQueue: Promise<void>;
         const subject = new Subject<T>();
         subject.cancelCallback = () => {
             const cancelInvocation: CancelInvocationMessage = this.createCancelInvocation(invocationDescriptor.invocationId);
 
             delete this.callbacks[invocationDescriptor.invocationId];
 
-            return this.sendWithProtocol(cancelInvocation);
+            return promiseQueue.then(() => {
+                return this.sendWithProtocol(cancelInvocation);
```

Comment:
I assume this failed before if you canceled the stream too early. Do we have any regression tests the promiseQueue part of this change?

Reply:
The promiseQueue doesn't change with this PR, I just use it somewhere I didn't before.
The test change needed was https://github.com/aspnet/AspNetCore/pull/7224/files#diff-e53c0bba1752f4416da08f9f40b4ac92R1078 to wait for a message since it now goes async.

Reply:
I know that the promiseQueue existed before. Are you saying that without this change, the "automatically sends multiple pings" would fail? Do you think that's a reliable regression test for this?

Reply:
The "stream can be canceled" test would fail because TS isn't threaded, so the cancel message wouldn't be sent before getting to the `expect(connection.sentData.length).toBe(3);` check. Now the cancel callback waits until the stream has finished being started before allowing the cancel to be sent.

I think the test is good.